### ISSUE TITLE
Lifecycle issue

### ISSUE THREAD

**Issue by PascalLiebl** [bluelinelabs/Conductor#263]
Hi,

I'm currently working on the following scenario: 
My MainActivity layout contains a ChangeHandlerFrameLayout and a regular button. By clicking the button, I want to alternate between two different controllers that inflate the ChangeHandlerFrameLayout by using the provided setBackstack(...) method.

...
switchButton.setOnClickListener(new View.OnClickListener() {
      @Override
      public void onClick(View v) {
            List<RouterTransaction<a>> routerTransactions = mRouter.getBackstack();
            Collections.reverse(routerTransactions);
            mRouter.setBackstack(routerTransactions, new FadeChangeHandler());
      }
});
...

So far everything works as expected, but there is a problem that occurs if the button is double-clicked (both views will be destroyed and the empty ChangeHandlerFrameLayout is displayed). If currently the first controller is inflated and I double-click the switchButton, I'm assuming that the same exact controller will still be visible after the ChangeHandler-animations are done.
The lifecycle of the first controller does not seem to be correct, because onCreateView() and onAttach() will not get called after the view of the second controller is detached and destroyed. 

I/Lifecycle: SecondController: onCreateView()
I/Lifecycle: SecondController: onAttach()
I/Lifecycle: FirstController: onDetach()
I/Lifecycle: FirstController: onDestroyView()
I/Lifecycle: SecondController: onDetach()
I/Lifecycle: SecondController: onDestroyView()

I have created a small example application to illustrate this behavior - what am I missing?

Thanks!
[ConductorDemo.zip](https://github.com/bluelinelabs/Conductor/files/910399/ConductorDemo.zip)

**Comment by stephanschuster** [comment]
My understanding of the problem is as follows:

Conductor's ChangeHandlerFrameLayout is intended to prevent such situations where clicks should be prevented while a transition is still ongoing. However, the way ChangeHandlerFrameLayout is implemented, it only prevents/intercepts event that are "inside" the layout. In your case the "Switch" button is outside the ChangeHandlerFrameLayout. So clicks on it will never be prevented.

The same issue arises in situations where you're triggering transitions (by pushing/popping controllers) on a child router (typically attached to a nested container/layout) by a button etc. which is outside the child layout. In my opinion this is a common scenario.

A solution to this problem could be a ChangeHandlerFrameLayout that delegates to parent ChangeHandlerFrameLayouts in order to intercept clicks on buttons located outside the child layout.

Here's my current implementation which works as expected:

`public class ChangeHandlerFrameLayout extends FrameLayout implements ControllerChangeListener {

    private int mInProgressTransactionCount;

    public ChangeHandlerFrameLayout(@NonNull Context context) {
        this(context, null);
    }

    public ChangeHandlerFrameLayout(@NonNull Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public ChangeHandlerFrameLayout(@NonNull Context context, @Nullable AttributeSet attrs, @AttrRes int defStyleAttr) {
        this(context, attrs, defStyleAttr, 0);
    }

    public ChangeHandlerFrameLayout(@NonNull Context context, @Nullable AttributeSet attrs, @AttrRes int defStyleAttr, @StyleRes int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
    }

    @Override
    public boolean onInterceptTouchEvent(MotionEvent event) {
        return mInProgressTransactionCount > 0 || super.onInterceptTouchEvent(event);
    }

    @Override
    public void onChangeStarted(@Nullable Controller to, @Nullable Controller from, boolean isPush, @NonNull ViewGroup container, @NonNull ControllerChangeHandler handler) {
        mInProgressTransactionCount++;

        ViewParent parent = getParent();
        while (parent != null) {
            if (parent instanceof ControllerChangeListener) {
                ((ControllerChangeListener) parent).onChangeStarted(to, from, isPush, container, handler);
                break;
            }
            parent = parent.getParent();
        }
    }

    @Override
    public void onChangeCompleted(@Nullable Controller to, @Nullable Controller from, boolean isPush, @NonNull ViewGroup container, @NonNull ControllerChangeHandler handler) {
        ViewParent parent = getParent();
        while (parent != null) {
            if (parent instanceof ControllerChangeListener) {
                ((ControllerChangeListener) parent).onChangeCompleted(to, from, isPush, container, handler);
                break;
            }
            parent = parent.getParent();
        }

        mInProgressTransactionCount--;

        if (mInProgressTransactionCount < 0) {
            mInProgressTransactionCount = 0;
        }
    }

}
`
A useful addition I'll probably make in the future is that the event which are intercepted will not just be swallowed but cached and dispatched later when change is completed.

**Comment by stephanschuster** [comment]
Okay, sorry, after a few tests with more complex applications it turned out that a simple counter is not enough. Here is the updated ChangeHandlerFrameLayout I currently use:

```
public class ChangeHandlerFrameLayout extends FrameLayout implements ControllerChangeListener {

    private List<Change> mChangesInProgress;

    public ChangeHandlerFrameLayout(@NonNull Context context) {
        this(context, null);
    }

    public ChangeHandlerFrameLayout(@NonNull Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public ChangeHandlerFrameLayout(@NonNull Context context, @Nullable AttributeSet attrs, @AttrRes int defStyleAttr) {
        this(context, attrs, defStyleAttr, 0);
    }

    public ChangeHandlerFrameLayout(@NonNull Context context, @Nullable AttributeSet attrs, @AttrRes int defStyleAttr, @StyleRes int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
    }

    @Override
    public boolean onInterceptTouchEvent(MotionEvent event) {
        return hasChangesInProgress() || super.onInterceptTouchEvent(event);
    }

    @Override
    public void onChangeStarted(@Nullable Controller to, @Nullable Controller from, boolean isPush, @NonNull ViewGroup container, @NonNull ControllerChangeHandler handler) {
        getChangesInProgress().add(new Change(to, from));

        ViewParent parent = getParent();
        while (parent != null) {
            if (parent instanceof ControllerChangeListener) {
                ((ControllerChangeListener) parent).onChangeStarted(to, from, isPush, container, handler);
                break;
            }
            parent = parent.getParent();
        }
    }

    @Override
    public void onChangeCompleted(@Nullable Controller to, @Nullable Controller from, boolean isPush, @NonNull ViewGroup container, @NonNull ControllerChangeHandler handler) {
        ViewParent parent = getParent();
        while (parent != null) {
            if (parent instanceof ControllerChangeListener) {
                ((ControllerChangeListener) parent).onChangeCompleted(to, from, isPush, container, handler);
                break;
            }
            parent = parent.getParent();
        }

        getChangesInProgress().remove(new Change(to, from));
    }

    private boolean hasChangesInProgress() {
        return mChangesInProgress != null && !mChangesInProgress.isEmpty();
    }

    private List<Change> getChangesInProgress() {
        if (mChangesInProgress == null) {
            mChangesInProgress = new ArrayList<>();
        }
        return mChangesInProgress;
    }

    private static class Change {

        private final String mToId;
        private final String mFromId;

        public Change(Controller to, Controller from) {
            mToId = to != null ? to.getInstanceId() : null;
            mFromId = from != null ? from.getInstanceId() : null;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (!(o instanceof Change)) return false;

            Change change = (Change) o;

            if (mToId != null ? !mToId.equals(change.mToId) : change.mToId != null) return false;
            if (mFromId != null ? !mFromId.equals(change.mFromId) : change.mFromId != null) return false;

            return true;
        }

        @Override
        public int hashCode() {
            int result = mToId != null ? mToId.hashCode() : 0;
            result = 31 * result + (mFromId != null ? mFromId.hashCode() : 0);
            return result;
        }

    }

}
```

@EricKuck: Do you have a better idea for dealing with the problem? Or could you even come up with a solution that does not even require this or the native ChangeHandlerFrameLayout? In a perfect world Conductor would be able to accept and gracefully handle user clicks during ongoing transactions without having blank screens in the end. So far I haven't look at the sources in detail to find out the actual reason for blank screens. But I'm pretty sure that there must be a better solution. Maybe you could comment on that. Thank you.

**Comment by EricKuck** [comment]
This issue should be fixed in the upcoming release. If you're still seeing problems, please feel free to reopen!

**Comment by stephanschuster** [comment]
Thank you for getting back to us.

I just upgraded to 2.1.3, replaced all my custom ChangeHandlerFrameLayouts (shown above) with plain FrameLayouts and immediately experienced blank screens again. So I would say this issue is not fixed yet. Sorry.

**Comment by EricKuck** [comment]
Yeah it looks like this was actually a separate issue than what was fixed in 2.1.3. That fix was for views which didn't attach when they should have. In this issue views actually are attached, but left with an alpha of zero due to multiple FadeChangeHandlers being executed simultaneously. This issue is now fixed as well in the latest snapshot (tested on the sample application posted in the first comment).

**Comment by stephanschuster** [comment]
Works. Perfect. Thank you, Eric.
